src/asl.ts

```diff
@@ -2129,33 +2319,152 @@ export class ASL {
+              return {
+                condition:
+                  expr.op === "&&"
+                    ? ASL.and(leftOutput.condition, right.condition)
+                    : expr.op === "||"
+                    ? ASL.or(leftOutput.condition, right.condition)
+                    : // ??: a boolean cannot be undefined or null, return the left side.
+                      leftOutput.condition,
+              };
+            }
+          }
+
+          const condition =
+            /**
+             * Left is a condition
+             * ??: checks for null or undefined, a condition will always be true or false
+             * &&: the right will be run when the result is true
+             * ||: the right will be run when the result is false
+             */
+            ASLGraph.isConditionOutput(leftOutput)
+              ? expr.op === "??"
+                ? ASL.trueCondition()
+                : expr.op === "&&"
+                ? leftOutput.condition
+                : ASL.not(leftOutput.condition)
+              : /**
+               * Left is a a json path
+               * ??: check for undefined or null
+               * &&: the right will be run when the result is truthy
+               * ||: the right will be run when the result is falsy
+               */
+              expr.op === "||"
+              ? ASL.isTruthy(leftOutput.jsonPath)
+              : expr.op === "&&"
+              ? ASL.not(ASL.isTruthy(leftOutput.jsonPath))
+              : ASL.isPresentAndNotNull(leftOutput.jsonPath);
+
+          return this.conditionState(expr, condition, leftOutput, right);
+        });
+      } else if (expr.op === "??=" || expr.op === "&&=" || expr.op === "||=") {
+        if (!isVariableReference(expr.left)) {
+          throw new SynthError(
+            ErrorCodes.Unexpected_Error,
+            "Expected left side of assignment to be a variable."
+          );
+        }
+        return this.evalExpr(expr.left, (left) => {
+          const right = this.eval(expr.right);
+
+          if (!ASLGraph.isJsonPath(left)) {
+            throw new SynthError(
+              ErrorCodes.Unexpected_Error,
+              `Expected assignment to target a variable, found: ${
+                ASLGraph.isLiteralValue(left) ? left.value : "boolean"
+              }`
+            );
+          }
+
+          /**
+           * &&: left ? right : left
+           * ||: left ? left : right
+           * ??: left !== undefined && left !== null ? left : right
+           */
+          const condition =
+            expr.op === "||="
+              ? ASL.isTruthy(left.jsonPath)
+              : expr.op === "&&="
+              ? ASL.not(ASL.isTruthy(left.jsonPath))
+              : ASL.isPresentAndNotNull(left.jsonPath);
+
+          return this.conditionState(
+            expr,
+            condition,
+            left,
+            right,
+            left.jsonPath
+          );
+        });
+      }
+      assertNever(expr.op);
```

Comment:
nit: I prefer to do `return assertNever(expr.op)`. I think the dangling `assertNever` is awkward.